// devextreme-vue/core/vue-helper.js
/*!
 * devextreme-vue
 * Version: 20.2.4
 * Build date: Tue Dec 01 2020
 *
 * Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-vue
 */

"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigurationOptions = exports.getNodeTypeOfComponent = exports.getNodeOptions = exports.saveComponentInstance = exports.usedConfigurationProps = exports.getComponentProps = exports.mount = exports.defaultSlots = exports.declaredTemplates = exports.configurationTemplate = exports.configurationDefaultTemplate = exports.configurationChildren = exports.getNormalizedProps = exports.getComponentInstance = exports.getComponentInfo = exports.getChildren = void 0;
var vue_1 = require("vue");
var helpers_1 = require("./helpers");
var children_processing_1 = require("./children-processing");
function getChildren(component) {
    if (!hasChildren(component) || !component.$_config) {
        return [];
    }
    var children = component.$.subTree && component.$.subTree.children;
    if (!Array.isArray(children)) {
        return [];
    }
    return children.filter(function (child) {
        if (!children_processing_1.isFragment(child)) {
            return child;
        }
        return;
    });
}
exports.getChildren = getChildren;
function getComponentInfo(component) {
    return getConfigurationOptions(component);
}
exports.getComponentInfo = getComponentInfo;
function getComponentInstance(component) {
    return component.type && component.type.$_componentInstance;
}
exports.getComponentInstance = getComponentInstance;
function getNormalizedProps(props) {
    var result = {};
    for (var propName in props) {
        if (props.hasOwnProperty(propName)) {
            result[helpers_1.camelize(propName)] = props[propName];
        }
    }
    return result;
}
exports.getNormalizedProps = getNormalizedProps;
function configurationChildren(component) {
    if (!component.children || !component.children.default) {
        return [];
    }
    return findConfigurationComponents(component.children.default());
}
exports.configurationChildren = configurationChildren;
function configurationDefaultTemplate(node) {
    if (!node.children || node.children === "object" || !node.children.default) {
        return;
    }
    return hasInlineTemplate(node.children.default()) ? node.children.default : undefined;
}
exports.configurationDefaultTemplate = configurationDefaultTemplate;
function configurationTemplate(node) {
    return configurationDefaultTemplate(node);
}
exports.configurationTemplate = configurationTemplate;
function declaredTemplates(component) {
    return component.$slots;
}
exports.declaredTemplates = declaredTemplates;
function defaultSlots(component) {
    var templates = declaredTemplates(component);
    if (!templates.default) {
        return [];
    }
    return templates.default();
}
exports.defaultSlots = defaultSlots;
function mount(options, parent, el) {
    var template = vue_1.createApp(options);
    template.provide("eventBus", parent.eventBus);
    template._context.components = Object.assign(parent.$.appContext.components, template._context.components);
    template._context.provides = Object.assign(parent.$.appContext.provides, template._context.provides);
    template._context.config = parent.$.appContext.config;
    template._context.directives = parent.$.appContext.directives;
    template._context.mixins = parent.$.appContext.mixins;
    return template.mount(el);
}
exports.mount = mount;
function getComponentProps(component) {
    var props = component.$.vnode.props || {};
    return getNormalizedProps(props);
}
exports.getComponentProps = getComponentProps;
function usedConfigurationProps(node) {
    return node.props;
}
exports.usedConfigurationProps = usedConfigurationProps;
function saveComponentInstance(component) {
    var nodeOptions = getNodeTypeOfComponent(component);
    if (nodeOptions) {
        nodeOptions.$_componentInstance = component;
    }
}
exports.saveComponentInstance = saveComponentInstance;
function getNodeOptions(component) {
    if (component.$) {
        return component.$.vnode;
    }
    return component;
}
exports.getNodeOptions = getNodeOptions;
function getNodeTypeOfComponent(component) {
    return component.$.vnode.type;
}
exports.getNodeTypeOfComponent = getNodeTypeOfComponent;
function findConfigurationComponents(children) {
    return children.filter(function (child) {
        if (children_processing_1.isFragment(child)) {
            return findConfigurationComponents(child.children || []);
        }
        if (child.type && typeof child.type === "object") {
            delete child.$_config;
            delete child.$_innerChanges;
            return child;
        }
        return;
    });
}
function hasInlineTemplate(children) {
    var hasTemplate = false;
    children.forEach(function (child) {
        if (!isConfiguration(child) && !children_processing_1.isFragment(child)) {
            hasTemplate = true;
        }
    });
    return hasTemplate;
}
function isConfiguration(child) {
    return child.type && typeof child.type === "object" && child.type.$_optionName;
}
function getConfigurationOptions(node) {
    return node.type;
}
exports.getConfigurationOptions = getConfigurationOptions;
function hasChildren(component) {
    return component.$.vnode && component.$.vnode.children && component.$.vnode.children.default;
}
